## Import Libraries


In [28]:
import pandas as pd
import numpy as np

## Read and Explore Data

In [29]:
# navigate to file path
file_path = r"C:\Users\Jahmar\Desktop\stopwatch\beer_reviews.csv" 
# read in csv file
beer_df = pd.read_csv(file_path)



In [30]:
# explore first 5 rows
beer_df.head()

brewery_id             brewery_name  review_time  review_overall  \
0       10325          Vecchio Birraio   1234817823             1.5   
1       10325          Vecchio Birraio   1235915097             3.0   
2       10325          Vecchio Birraio   1235916604             3.0   
3       10325          Vecchio Birraio   1234725145             3.0   
4        1075  Caldera Brewing Company   1293735206             4.0   

   review_aroma  review_appearance review_profilename  \
0           2.0                2.5            stcules   
1           2.5                3.0            stcules   
2           2.5                3.0            stcules   
3           3.0                3.5            stcules   
4           4.5                4.0     johnmichaelsen   

                       beer_style  review_palate  review_taste  \
0                      Hefeweizen            1.5           1.5   
1              English Strong Ale            3.0           3.0   
2          Foreign / Export Stout            3.0           3.0   
3                 German Pilsener            2.5           3.0   
4  American Double / Imperial IPA            4.0           4.5   

                beer_name  beer_abv  beer_beerid  
0            Sausa Weizen       5.0        47986  
1                Red Moon       6.2        48213  
2  Black Horse Black Beer       6.5        48215  
3              Sausa Pils       5.0        47969  
4           Cauldron DIPA       7.7        64883

In [31]:
# Get a list of all columns
beer_df.columns

Index(['brewery_id', 'brewery_name', 'review_time', 'review_overall',
       'review_aroma', 'review_appearance', 'review_profilename', 'beer_style',
       'review_palate', 'review_taste', 'beer_name', 'beer_abv',
       'beer_beerid'],
      dtype='object')

In [5]:
# Check columns data type
beer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1586614 entries, 0 to 1586613
Data columns (total 13 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   brewery_id          1586614 non-null  int64  
 1   brewery_name        1586599 non-null  object 
 2   review_time         1586614 non-null  int64  
 3   review_overall      1586614 non-null  float64
 4   review_aroma        1586614 non-null  float64
 5   review_appearance   1586614 non-null  float64
 6   review_profilename  1586266 non-null  object 
 7   beer_style          1586614 non-null  object 
 8   review_palate       1586614 non-null  float64
 9   review_taste        1586614 non-null  float64
 10  beer_name           1586614 non-null  object 
 11  beer_abv            1518829 non-null  float64
 12  beer_beerid         1586614 non-null  int64  
dtypes: float64(6), int64(3), object(4)
memory usage: 157.4+ MB


## Which brewery produces the strongest beers by ABV%?

In [6]:
# Get the Max ABV Percentage to determine which is the strongest
strongabv = beer_df[beer_df.beer_abv == beer_df.beer_abv.max()]

In [7]:
# sort values by descending to show top 10 
Top_10_By_ABV = beer_df.sort_values("beer_abv",axis=0,ascending=False).head(10)
Top_10 = Top_10_By_ABV[['brewery_name','beer_abv']]
Top_10

brewery_name  beer_abv
12919   Schorschbräu      57.7
12939   Schorschbräu      43.0
12940   Schorschbräu      43.0
746385       BrewDog      41.0
746387       BrewDog      41.0
746386       BrewDog      41.0
746384       BrewDog      41.0
746358       BrewDog      41.0
746359       BrewDog      41.0
746360       BrewDog      41.0

##### Answer: Schorschbräu is the brewery that produces the strongest ABV% at 57.7%

##  If you had to pick 3 beers to recommend using only the data in this dataset, which would you pick?

If I was to pick 3 beers,  I'd go based on the mean of overall reviews, taste reviews, review appearance and beer abv. 


In [8]:
# create df and groupby beer name. Find the mean for overall, appearance, taste and beer abv
recommendation = beer_df.groupby('beer_name').agg({'review_overall': np.mean,
                                'review_appearance': np.mean,
                                'review_taste': np.mean,
                                'beer_abv': np.mean})

# Sort columns by descending order
top_n = recommendation.sort_values(['review_overall','review_appearance',
                                    'review_taste','beer_abv'], 
                        ascending=[False, False, False, False])
# Show top 3
top_n.head(3)

review_overall  review_appearance  \
beer_name                                                                     
Campfire Joe                                         5.0                5.0   
Old Incinerator Brandy Barrel Barleywine             5.0                5.0   
Chicago Terminal Oatmeal Bourbon Stout               5.0                5.0   

                                          review_taste  beer_abv  
beer_name                                                         
Campfire Joe                                       5.0     17.35  
Old Incinerator Brandy Barrel Barleywine           5.0     13.00  
Chicago Terminal Oatmeal Bourbon Stout             5.0     11.00

##### Answer: Top 3 Beers I would reccomend based on the popular mean in reviews 
      1. Campfire Joe
      2. Old Incinerator Brandy Barrel Barleywine
      3. Chicago Terminal Oatmeal Bourbon Stout	

## Which of the factors (aroma, taste, appearance, palette) are most important in determining the overall quality of a beer?
   

In [15]:
#Find a relationship between the overall, appearance, taste, and palette colmun 
beer_factors = beer_df

beer_factors = beer_df.corr()

beer_factors

brewery_id  review_time  review_overall  review_aroma  \
brewery_id           1.000000     0.220954       -0.009331      0.019069   
review_time          0.220954     1.000000        0.028724      0.067453   
review_overall      -0.009331     0.028724        1.000000      0.616013   
review_aroma         0.019069     0.067453        0.616013      1.000000   
review_appearance    0.001993     0.058233        0.501732      0.561029   
review_palate       -0.000499     0.057247        0.701914      0.616947   
review_taste        -0.002672     0.044498        0.789816      0.716776   
beer_abv             0.063731     0.144885        0.138512      0.332570   
beer_beerid          0.424097     0.461535        0.003794      0.077669   

                   review_appearance  review_palate  review_taste  beer_abv  \
brewery_id                  0.001993      -0.000499     -0.002672  0.063731   
review_time                 0.058233       0.057247      0.044498  0.144885   
review_overall              0.501732       0.701914      0.789816  0.138512   
review_aroma                0.561029       0.616947      0.716776  0.332570   
review_appearance           1.000000       0.566634      0.546980  0.263941   
review_palate               0.566634       1.000000      0.734135  0.286711   
review_taste                0.546980       0.734135      1.000000  0.290827   
beer_abv                    0.263941       0.286711      0.290827  1.000000   
beer_beerid                 0.054794       0.044510      0.030402  0.201818   

                   beer_beerid  
brewery_id            0.424097  
review_time           0.461535  
review_overall        0.003794  
review_aroma          0.077669  
review_appearance     0.054794  
review_palate         0.044510  
review_taste          0.030402  
beer_abv              0.201818  
beer_beerid           1.000000

##### Answer: In trying to determine the quality of a beer, taste is by far the most important factor in determining the quality of a beer as it has a .79 correlation with the overall reviews. 

## 4) If I typically enjoy a beer due to its aroma and appearance, which beer style should I try?

My approach to this will be to count the number of reviews for both columns, get the mean of reviews for both columns and get the max amount of reviews for both columns. This will help determine which is popular in both areas.

In [17]:
# Create df for the columns aroma, appearance and beer style.
beerstyle = beer_df[['review_aroma','review_appearance','beer_style']]
beerstyle.head()

review_aroma  review_appearance                      beer_style
0           2.0                2.5                      Hefeweizen
1           2.5                3.0              English Strong Ale
2           2.5                3.0          Foreign / Export Stout
3           3.0                3.5                 German Pilsener
4           4.5                4.0  American Double / Imperial IPA

In [25]:
# Get the mean, count and max for a aroma,appearance and groupby beerstyle

beerstyle1 = beer_df.groupby('beer_style').agg({'review_aroma': np.mean,
                                'review_appearance': np.mean})
beerstyle1 = pd.DataFrame(beerstyle1.reset_index())
beerstyle1.columns = ['beer_style','review_aroma_mean','review_appearance_mean']

beerstyle2 = beer_df.groupby('beer_style').agg({'review_aroma': np.max,
                                             'review_appearance': np.max})
beerstyle2 = pd.DataFrame(beerstyle2.reset_index())
beerstyle2.columns = ['beer_style','review_aroma_max','review_appearance_max']

beerstyle3 = beer_df.groupby('beer_style').agg({'review_aroma': np.size,
                                             'review_appearance': np.size})
beerstyle3 = pd.DataFrame(beerstyle3.reset_index())
beerstyle3.columns = ['beer_style','review_aroma_count','review_appearance_count']

beerstyle = pd.merge(pd.merge(beerstyle2, beerstyle1, on='beer_style'),beerstyle3, on='beer_style')

In [26]:
# sort columns in desc order and limit 3 for top 3
recommended = beerstyle.sort_values(['review_aroma_count','review_appearance_count','review_aroma_mean',
                                     'review_appearance_mean','review_aroma_max','review_appearance_max'], 
                                    ascending=[False, False,False, False, False, False])
top = recommended [:3]['beer_style']
top

12                      American IPA
9     American Double / Imperial IPA
14           American Pale Ale (APA)
Name: beer_style, dtype: object

##### The Top 3 Beerstyles I would reccomend are 
      1. American IPA
      2. American Double / Imperial IPA
      3. American Pale Ale (APA)